### ISSUE TITLE
Major privacy/freedom issue with TEKs upload procedure.

### ISSUE THREAD

**Issue by ivanvisconti** [immuni-app/immuni-documentation#27]
Citizens should freely decide to upload or not their TEKs once they discover to be infected. Moreover, their decision should remain private. As far as I understand (please correct me in case I'm wrong), the procedure proposed by Immuni violates this freedom/privacy. Indeed the infected citizen when asked by the Healthcare Operator might feel under a psychological pressure due to so called "civic sense" and could be afraid that the Healthcare Operator would then gossip about her/his decision. After accepting to do it, there is an OTP generated by the app and validated by the Healthcare Operator. If the user decides to abort (i.e., the user does not complete the upload), its decision would not remain private since the server will not receive the TEKs  from the specific citizen that corresponds to the OTP validated by the Healthcare Operator, and that knows the identity of the citizen. This is not acceptable. 

Currently there are many other things that I dislike of Immuni but I would like to at least propose a fix to this problem. After the validation of the OTP, the app should offer to the citizen the possibility to opt among two options: 1) upload of the real TEKs or 2) upload of randomly generated TEKs (that have never been sent as identifier beacons). In this way the citizen can freely decide not to send the real TEAKs and still this decision remains known uniquely to him. This is privacy by design (FYI).

Please let me know if my analysis is correct, and in the affirmative case please explain if you are going to update Immuni implementing the fix that I've just suggested (otherwise, explain why you're not protecting the privacy of the citizens).

Additionally, the citizen might want to prefer a mixed solution. Consider for example a citizen that 3 days before the test was in some risky situation so that communicating the TEK corresponding to that day might be harmful (e.g., the citizen could be fired), while there is no problem to communicate the TEKs corresponding to all other days. An even better solution, therefore, consists of offering to the citizen the possibility of selectively choosing the reak TEKs to upload, so that the discarded ones will be replaced by randomly chosen TEKs. Basically the citizen will have a form with the last X days (I don't know if X=14 or some other number) and day by day the citizen will select or not a checkbox in order to send the real TEK or the fake TEK. From a usability point of view I think the user should have 3 options: 1) upload all real; 2) upload all fake; 3) mixed upload. Only in case the 3rd option is selected, then the citizen should receive a form with a day-by-day checkbox. It's certainly beneficial that the citizen can perform this action far from the Healthcare Operator so that the time spent to choose the favorite option does not leak any information on her/his decision.

**Comment by codker** [comment]
I have the same concern, this is unacceptable for the privacy, it’s a big mistake.

The process must be inverted, it’s the operator that should give an OTP to the infected person that inputs it in the app if he wants to.

The operator shouldn’t be concerned if the user accepts to share it’s data or not, this is a big red flag for people’s privacy.

**Comment by GennAvi** [comment]
If you want to keep the current authorization mechanism and apply the fix proposed by @ivanvisconti , keep also in mind that the overall procedure must still be handled properly at UI level. Maybe the validation of the code is done via telephone call, since the user may be in precautionary quarantine. If this is always the case, then the UI that is shown in product description may just need an additional button. 
But if this validation is done in presence of the health operator, the UI must enforce the privacy of the user in some way (e.g., by freezing for an appropriate amount of time, for example 30 minutes, after the validation of the code is performed). 
By doing this, the privacy of the user would be enforced by design. In fact, if this mitigation is not in place, right after the validation of the code the upload section could appear and the presence of the operator might push the infected user in rushing to make a decision in front of him, and maybe reporting data even if he does not want, only because the operator is watching him.
Let me know if you want this in a separate issue to keep track of it more easily.

**Comment by mmeloni** [comment]
In my opinion the key of the infected users need to be uploaded by third party machine/app managed by health care operator. In this way there is no correlation between the key and the infected.

**Comment by tommyazz** [comment]
The thing is that a device produces a local set of random keys to be
incorporated in BLE beacons that are broadcasted to other users. If a
certain user tests positive to Covid, those keys, locally present in your
device, need to be fetched by other users and compared with their received
keys in the past 14 days. Therefore they have to be upload by the user who
tested positive. I don't see any other solution.

**Comment by tommyazz** [comment]
@ivanvisconti Let's adopt the second solution you propose (i.e. to upload randomly generated TEKs - never seen before). 
Before that, I would like to discuss the Apple-Google framework. Basically every 24 hours a new local TEK is generated. This TEK is then used to derive RPIs (Rolling proximity identifiers). Each RPI is sent through the BLE protocol and collected by other users (so first thing, TEKs are not sent). 
After a person tests positive, locally generated TEKs are upload to the server and fetched by other users. The fetched TEKs are used to retrieve the RPIs produced by those TEKs, then these RPIs are going to be compared with the RPIs received by the user (a person who could have been infected by the person who tested positive) in the past days. If there's a match then the user is notified otherwise nothing happens. 
Given this, I think it's pretty clear that, in order for other users to compare their own received RPIs, the original TEKs are required to be fetched. The upload of randomly generated TEKs will result in discrepancies between the RPIs previously generated with a different TEK (received by other users) and the RPIs associated with the upload TEKs.

**Comment by ivanvisconti** [comment]
@tommyazz you talk about a second solution but I've proposed only one solution. Also, let me stress that the upload of randomly generated TEKs should be an option offered to the infected citizen in order to protect her/his right not to communicate her/his health situation to whoever has tracked her/him in the previous days, while at the same time keeping private this decision.

**Comment by tommyazz** [comment]
@ivanvisconti Yeah, second option of the solution, sorry.
Well, you're not exactly communicating your own health situation to others. Given that the others don't know precisely who has been tested positive but they just know they have been in contact with someone who tested positive.
The best solution, if someone is concerned about their privacy, is to avoid the use of this APP, since it's not mandatory. The upload for random TEKs is useless from the point of view of how the platform technically works.

**Comment by ivanvisconti** [comment]
> The best solution, if someone is concerned about their privacy, is to avoid the use of this APP, since it's not mandatory. The upload for random TEKs is useless from the point of view of how the platform technically works.

No, I strongly disagree. I think a citizen should have a right to use the app, to get notification and then to object on uploading her/his real TEKs depending on what happened in the last 14 days. The option to use real or fake ID is fundamental in order to protect the privacy of the citizen taking this decision to upload or not real TEKs. If you still are in trouble in understanding this fundamental right for citizens, consider for example the analogue case of elections where a citizen decides to go to vote (otherwise it would be publicly known that he/she did not vote) and then submits an empty ballot (but nobody knows that he/she did it, so the privacy of its choice is preserved). I hope this helps to clarify.

**Comment by tommyazz** [comment]
@ivanvisconti I already understood it, no worries! 
I still don't find your proposal useful since a user can decide whether to upload its data if he/she is found positive. And even if someone, who is found positive, does not upload its data, he/she can still use the app. The Health System doesn't know if you are using the app or not, even if you are tested positive. 
The voting system is a bit different since everyone is identified through the voting card and their name is screamed at the act of voting. The TEKs, generated by a device, don't identify the person who is using the APP. Maybe this clarifies my position.
![Screenshot 2020-05-17 at 16 21 24](https://user-images.githubusercontent.com/26226831/82151228-c783be80-985a-11ea-8f93-c5e989380af5.jpg)

**Comment by ivanvisconti** [comment]
@tommyazz I disagree and actually I feel that I can say  that you are wrong. A person that is found positive is not anonymous, indeed the Health Operator knows her/his name (as far as I know in Italy there is no anonymous procedure to get a Covid-19 test). Then there is a procedure with the activation code to run with the Health Operator (not simply something that the citizen can do by herself/himself). I don't see how one can freely decide preserving the privacy of her/his decision with the current description of Immuni. The fix that I've proposed allows citizens to freely take a decision without compromising the privacy of their decision (the two things are actually connected because a lack of privacy impacts on freedom).

**Comment by codker** [comment]
@tommyazz, sorry but you stated 

>  The Health System doesn't know if you are using the app or not, even if you are tested positive.

This is not correct as the Health Operator must allow the positive person to upload it's keys through an OTP, so he must know if he is using the app or not and if he wants to share the positivity or not.

That, for me, is wrong.

As I already said, Immuni should reverse the process:

- a person is tested positive
- the Health Operator gives him an (anonymous) OTP that will expire in 24h (not the other way around)
- the infected person has 24h to take a decision and input the OTP in the Immuni app, on its own

This will respect the privacy as the Health Operator would never know if the person has the app installed and if he wants to share the exam result.

Maybe I'm missing something about the actual process?

**Comment by GennAvi** [comment]
@liberatiluca This is not enough, it just solves a local problem, if the same auth code which is given to the citizen is also forwarded, along with his real identity, to the server (i.e. the government), the server would be able to determine if the infected citizen uploaded the data or not. Therefore, the fix proposed by @ivanvisconti would still be needed.

**Comment by codker** [comment]
Yeah absolutely, I was just talking about the user’s privacy vs the Health Operator job.

That is needed too.

**Comment by tommyazz** [comment]
> @tommyazz I disagree and actually I feel that I can say that you are wrong. A person that is found positive is not anonymous, indeed the Health Operator knows her/his name (as far as I know in Italy there is no anonymous procedure to get a Covid-19 test). Then there is a procedure with the activation code to run with the Health Operator (not simply something that the citizen can do by herself/himself). I don't see how one can freely decide preserving the privacy of her/his decision with the current description of Immuni. The fix that I've proposed allows citizens to freely take a decision without compromising the privacy of their decision (the two things are actually connected because a lack of privacy impacts on freedom).

I think it was clear I was talking from the application point of view. Of course, If I go to the hospital for a COVID test, they know who I am (through my SSN). There's no need to clear this out. 
The OTP procedure starts when a user, who tested positive, willingly DECIDES to upload his/her epidemiological data to the server. 

> the citizen can freely decide not to send the real TEAKs and still this decision remains known uniquely to him

The citizen can freely decide not to start the upload of data (which, in turn, starts the OTP procedure) and this decision remains known uniquely to him.

To be clear, privacy and "freedom" matters concern me too. I simply don't think this proposal adds something useful. Maybe we should look at other critical parts.

Before saying all of this, I went through the entire Immuni's documentation (which I think it's pretty clear) and the one provided by Apple and Google.

**Comment by tommyazz** [comment]
> @tommyazz, sorry but you stated
> 
> > The Health System doesn't know if you are using the app or not, even if you are tested positive.
> 
> This is not correct as the Health Operator must allow the positive person to upload it's keys through an OTP, so he must know if he is using the app or not and if he wants to share the positivity or not.

How does the Health Operator know this if I don't start the Upload procedure?

> 
> That, for me, is wrong.
> 
> As I already said, Immuni should reverse the process:
> 
> * a person is tested positive
> * the Health Operator gives him an (anonymous) OTP that will expire in 24h (not the other way around)
> * the infected person has 24h to take a decision and input the OTP in the Immuni app, on its own
> 
> This will respect the privacy as the Health Operator would never know if the person has the app installed and if he wants to share the exam result.

Probably this procedure has its good points, but, If you upload your data that means you have the app. If you don't decide to start the upload, how do they know if you have the app?
Since the app is not mandatory, even if you test positive you are not required to upload your data. Therefore, why should the authorities even blame you?  

> 
> Maybe I'm missing something about the actual process?

**Comment by codker** [comment]
@tommyazz I see your point, but...with the current process, if a person is tested positive, an operator MUST ask him if he installed the app and wants to upload the keys, because the person MUST start the upload process to give an OTP to the operator.

If the process is reversed, the operator will just give an OTP to every positive tested person saying: “if you have the app you can use this code to help other people you came in contact”. That’s it. The operator will never know if the person installed the app and wants to upload the keys.

It’s just a small detail in the bigger picture, but I think it’s important.

I hope to have explained my point clearly now.

**Comment by tommyazz** [comment]
> @tommyazz I see your point, but...with the current process, if a person is tested positive, an operator MUST ask him if he installed the app and wants to upload the keys, because the person MUST start the upload process to give an OTP to the operator.

Why the operator must ask? Shouldn't be more like "may I remind you that there is this app..."? Is this explicitly stated somewhere?
I think the upload can be started any time a user wants to. 

> 
> If the process is reversed, the operator will just give an OTP to every positive tested person saying: “if you have the app you can use this code to help other people you came in contact”. That’s it. The operator will never know if the person installed the app and wants to upload the keys.

It might be a good strategy.  The operator will know you have the app when you use that OTP. Otherwise, it doesn't have this info. Does the operator know when an OTP is used and to whom is associated?

> 
> It’s just a small detail in the bigger picture, but I think it’s important.
> 
> I hope to have explained my point clearly now.
> 

What happens if someone gives its OTP to an attacker or if an attacker stoles an OTP. The attacker might be able to upload fake data, which causes wrong alarms and waste of health resources. So, probably, a validation of the operator is always needed.

**Comment by ivanvisconti** [comment]
@tommyazz,

you say: "The citizen can freely decide not to start the upload of data (which, in turn, starts the OTP procedure) and this decision remains known uniquely to him."

This is wrong. If the code is validated and the Health Operator knows the identity of the citizen, then the server+HO will register the missing TEKs from that citizen, and they will conclude that he/she did not upload data. The citizen should be free to take her/his decision privately, without others being able to detect it.

you say: "How does the Health Operator know this if I don't start the Upload procedure?"

HO and server are both controlled by the government a citizen might be afraid of collusion. 

you say: "If you don't decide to start the upload, how do they know if you have the app?"

HO asks the citizen about it and will figure it out from the answer.

you say: " why should the authorities even blame you?"

It seems that you're commenting an issue without evening reading the content of the issue (which is very weird). It's clearly stated in the text of the issue: <<might feel under a psychological pressure due to so called "civic sense" and could be afraid that the Healthcare Operator would then gossip about her/his decision>>. So it's not a matter of the authorities blaming the citizen, it's a matter of the citizen feeling under pressure. Do you care about such people or perhaps you prefer to suppress their rights?

I hope you support the idea of people being free to choose what they want to do without psychological pressure, keeping private their decisions.

**Comment by ivanvisconti** [comment]
@tommyazz 

> Why the operator must ask? Shouldn't be more like "may I remind you that there is this app..."? Is this explicitly stated somewhere?

Can you exclude that an operator might ask? If no, then we must make the system resilient to this case. You can't design a system assuming that all players follow the rules, otherwise as soon as somebody deviates, the system will collapse. This is so straightforward, it's becoming embarrassing for me to write such things.

**Comment by codker** [comment]
@tommyazz It all comes down to what @ivanvisconti is saying.

The thing is that the whole process must be worked from the people perspective, not the government one.

>  The operator will know you have the app when you use that OTP. Otherwise, it doesn't have this info. Does the operator know when an OTP is used and to whom is associated?

The HO gives the OTP to the user and stop there, his job is done, he will never know if someone had the app installed, has agreed to send the keys or used the OTP. If anyone can see that kind of info relative to a user, the system is not anonymous.

> What happens if someone gives its OTP to an attacker or if an attacker stoles an OTP

That's why I said 24 hours limit, that should mitigate almost any possible attack. If someone finds out that I'm positive and stole my OTP in 24 hours, I think we all have another big problem to think of.

That said, I understand you pov.

**Comment by tommyazz** [comment]
> @tommyazz,
> 
> you say: "The citizen can freely decide not to start the upload of data (which, in turn, starts the OTP procedure) and this decision remains known uniquely to him."
> 
> This is wrong. If the code is validated and the Health Operator knows the identity of the citizen, then the server+HO will register the missing TEKs from that citizen, and they will conclude that he/she did not upload data. The citizen should be free to take her/his decision privately, without others being able to detect it.

Which missing TEKs? The user has just started the process to upload its TEKs? 
Oh, I see, so this is basically the case when someone decides to upload its data. The procedure is started, the OTP is validated by the healthcare operator. But suddenly the user decides not to complete the procedure. Well, why did you start it? You could avoid this by not starting it. 
The user has the time to make the best decision before starting the upload process.
This means what I stated was not entirely wrong!  

> 
> you say: "How does the Health Operator know this if I don't start the Upload procedure?"
> 
> HO and server are both controlled by the government a citizen might be afraid of collusion.

This doesn't answer the question.

> 
> you say: "If you don't decide to start the upload, how do they know if you have the app?"
> 
> The HO asks the citizen about it and will figure it out from the answer.

How can the HO be so sure by asking? Do they run a polygraph (just kidding)?
I imagine the uploading procedure can be started by the user "at home" not necessarily at the hospital.
And even if they ask, is there going to be a "Has the APP/Has not the APP" flag in our health profiles? 

> 
> you say: " why should the authorities even blame you?"
> 
> It seems that you're commenting an issue without evening reading the content of the issue (which is very weird). It's clearly stated in the text of the issue: <<might feel under a psychological pressure due to so called "civic sense" and could be afraid that the Healthcare Operator would then gossip about her/his decision>>. So it's not a matter of the authorities blaming the citizen, it's a matter of the citizen feeling under pressure. Do you care about such people or perhaps you prefer to suppress their rights?

It seems to me you are creating an issue without even looking and analyzing the documentation of this app and the documentation provided by A-G. 
Back to your proposal, where is the civic sense if I'm tested positive and I upload random useless TEKs (option 2)? 
Is it just a matter of rights or also a matter of duties?
Do you care about the people who can suffer bad consequences because they are not informed?
I am reversing the question because, I think, also the opposite should be considered.

> 
> I hope you support the idea of people free to choose what they want to do without psychological pressure, keeping private their decisions.

Yes, I support it!  But, I would rather feel an ill-defined pressure from my HO than the pressure of having uploaded random TEKs, which resulted in possible sick people being not promptly informed. You clearly stated that it is a matter of the citizen feeling under pressure.

**Comment by tommyazz** [comment]
> @tommyazz It all comes down to what @ivanvisconti is saying.
> 
> The thing is that the whole process must be worked from the people perspective, not the government one.
> 
> > The operator will know you have the app when you use that OTP. Otherwise, it doesn't have this info. Does the operator know when an OTP is used and to whom is associated?
> 
> The HO gives the OTP to the user and stop there, his job is done, he will never know is someone had the app installed, has agreed to send the keys or used the OTP. If anyone can see that kind of info relative to a user, the system is not anonymous.
> 
> > What happens if someone gives its OTP to an attacker or if an attacker stoles an OTP
> 
> That's why I said 24 hours limit, that should mitigate almost any possible attack. If someone finds out that I'm positive and stole my OTP in 24 hours, I think we all have another big problem to think of.
> 
> That said, I understand you pov.

I understand your POV too, and I accept any criticism.
Sorry, but I'd say that 24 hours is plenty of time. In any case, what you propose can be investigated. And I think it could be a better solution than the one proposed with this issue.
Thanks for your comments!

**Comment by tommyazz** [comment]
> Can you exclude that an operator might ask? If no, then we must make the system resilient to this case. You can't design a system assuming that all players follow the rules, otherwise as soon as somebody deviates, the system will collapse. This is so straightforward, it's becoming embarrassing for me to write such things.

That said, you can't even design a system because someone could feel an ill-defined pressure. 
Listen, if you open an issue but you are not open to any criticism just tell it. I will not comment any further.

**Comment by ivanvisconti** [comment]
> > @tommyazz,
> > you say: "The citizen can freely decide not to start the upload of data (which, in turn, starts the OTP procedure) and this decision remains known uniquely to him."
> > This is wrong. If the code is validated and the Health Operator knows the identity of the citizen, then the server+HO will register the missing TEKs from that citizen, and they will conclude that he/she did not upload data. The citizen should be free to take her/his decision privately, without others being able to detect it.
> 
> Which missing TEKs? The user has just started the process to upload its TEKs?
> Oh, I see, so this is basically the case when someone decides to upload its data. The procedure is started, the OTP is validated by the healthcare operator. But suddenly the user decides not to complete the procedure. Well, why did you start it? You could avoid this by not starting it.
> The user has the time to make the best decision before starting the upload process.
> This means what I stated was not entirely wrong!
> 

It is entirely wrong, because the decision to start the upload process could be taken under psychological pressure. This is written in my issue and repeated already several times in previous answers. It corresponds for instance to clients that buy more when entering in a store under the pressure of the sale assistant. Here we are discussing about very sensitive data touching civil rights and we should make sure to provide the technology allowing citizens to take free decisions. 

It's not worthy to repeat 10 more times this concept, if you don't care about protecting such people then we just disagree. That's it. I stand with the subject of the issue, there is a major privacy/freedom issue with TEKs upload procedure. I've proposed an easy fix. There is so far no valid reason to ignore the fix, unless one would really like to limit freedom and civil rights.

**Comment by tommyazz** [comment]
> > > @tommyazz,
> > > you say: "The citizen can freely decide not to start the upload of data (which, in turn, starts the OTP procedure) and this decision remains known uniquely to him."
> > > This is wrong. If the code is validated and the Health Operator knows the identity of the citizen, then the server+HO will register the missing TEKs from that citizen, and they will conclude that he/she did not upload data. The citizen should be free to take her/his decision privately, without others being able to detect it.
> > 
> > 
> > Which missing TEKs? The user has just started the process to upload its TEKs?
> > Oh, I see, so this is basically the case when someone decides to upload its data. The procedure is started, the OTP is validated by the healthcare operator. But suddenly the user decides not to complete the procedure. Well, why did you start it? You could avoid this by not starting it.
> > The user has the time to make the best decision before starting the upload process.
> > This means what I stated was not entirely wrong!
> 
> It is entirely wrong, because the decision to start the upload process could be taken under psychological pressure. This is written in my issue and repeated already several times in previous answers. It corresponds for instance to clients that buy more when entering in a store under the pressure of the sale assistant. Here we are discussing about very sensitive data touching civil rights and we should make sure to provide the technology allowing citizens to take free decisions.
> 
> It's not worthy to repeat 10 more times this concept, if you don't care about protecting such people then we just disagree. That's it. I stand with the subject of the issue, there is a major privacy/freedom issue with TEKs upload procedure. I've proposed an easy fix. There is so far no valid reason to ignore the fix, unless one would really like to limit freedom and civil rights.

I would avoid the use of expressions like "This is entirely wrong" or "This is not correct" when interacting with other people. You are literally NO ONE to treat others' opinions like garbage and it is also not polite. I don't think I've ever used such expressions in my previous comments, if yes I apologize as it is not my way to interact with people.

"I've proposed an easy fix. There is so far no valid reason to ignore the fix, unless one would really like to limit freedom and civil rights". 
If this is really an issue, I hope it'll be fixed and explained to all of us in a less arrogant way.

**Comment by gvdr** [comment]
> you can't even design a system because someone could feel an ill-defined pressure

I would invite you to take a look at how the GDPR defines free, informed, consent. Psychological pressures are taken into account, and they constitute a problem. Especially so in the scenario of this app.

**Comment by tommyazz** [comment]
> > you can't even design a system because someone could feel an ill-defined pressure
> 
> I would invite you to take a look at how the GDPR defines free, informed, consent. Psychological pressures are taken into account, and they constitute a problem. Especially so in the scenario of this app.

Yes, thanks for sharing! 
Let me clarify that I have never said that psychological pressures should not be taken into account. I was talking of something ill-defined, which is different. In particular, a solution that addresses a possible problem by creating possible others.

**Comment by gvdr** [comment]
@tommyazz, just to clarify: are you involved with the team or the task force in any way? I'm starting to lose track of who's involved and who's not.

**Comment by tommyazz** [comment]
@gvdr No I'm not involved with any of them.

**Comment by ivanvisconti** [comment]
@immuniopensource ...pronto-C6?

**Comment by ALBUCHERQUE** [comment]
I worked with various military grade application. Ive modified diffiehellman, keccak and aes. I reviewed the chips used by the pentagono end nasa. Their procedure is perfect and is the one most advanced today. Who are you to say that there is a problem?! Wait for the code and be mute.

**Comment by gvdr** [comment]
Ok, we have the troll guy.

**Comment by vi-enne** [comment]
@immuniopensource 
Since issue [#57](https://github.com/immuni-app/immuni-app-android/issues/57) in repo [immuni-app-android](https://github.com/immuni-app/immuni-app-android) has been closed referencing this issue, I would seize the opportunity and revitalize this issue.

Thanks,
Vittorio

**Comment by ivanvisconti** [comment]
> @immuniopensource
> Since issue [#57](https://github.com/immuni-app/immuni-app-android/issues/57) in repo [immuni-app-android](https://github.com/immuni-app/immuni-app-android) has been closed referencing this issue, I would seize the opportunity and revitalize this issue.
> 
> Thanks,
> Vittorio

Thanks. 11 days have passed already and we didn't get any feedback from the Immuni team. The same holds for issue #18. This is weird since in both cases I've proposed pretty trivial fixes. It's becoming hard not to suspect that they might have designed these procedures with the goal of building a database with the real identities of the infected citizens associated to all pseudonyms that they have used in the previous 2 weeks. Combining this database with the installation of beacons in strategic places that will (undetectably) scan the pseudonyms, there is quite a lot of surveillance that can be done. The risks of realizing Paparazzi, Orwell and Brutus attacks are becoming more and more concrete.

**Comment by GennAvi** [comment]
It seems that the new documentation unveiled a detail that makes the problem @ivanvisconti noticed even worse:

> The OTP has a time to live, after which it can no longer be used to authorise an upload. The time to live is set at 2 minutes and 30 seconds initially, but it can be adjusted later if required. 

This time has to be greatly extended, if you are concerned of OTP collisions you should just increase its length.
I think it is time for @immuniopensource to make explicit if they are going to address or not this problem and what is their vision about this issue, as well as issue 18. 
Regarding this issue, which is pretty full of noise, if you want a clear idea the first 3 comments are enough IMHO.

**Comment by immuniteam** [comment]
Hi @ivanvisconti , thanks for raising this issue. 
Things are slightly different from what you describe. A user that tests positive for the virus communicates their OTP by phone to the same Healthcare Operator that disclosed to them the positive outcome of the swab test. Since the conversation happens via phone, the user and the Healthcare Operator aren’t close to each other. There’s no way for the Healthcare Operator to communicate the user’s identity (which is known to them) to the Immuni server. For this reason Immuni’s server doesn’t know the identity of the positive user or of the Healthcare Operator and the latter doesn’t receive any acknowledgement that the upload was completed successfully, meaning that even if the user decides to abort the upload at any point in time, the Healthcare Operator won’t know and their decision would remain private. Finally, Healthcare Operators are instructed to inform the positive user that the upload procedure is not mandatory and that the user should freely decide if they want to proceed.
Given the above, we made the choice to keep the user experience extremely simple to minimise the chance that the presence of extra steps would cause the keys not to be uploaded when users want to.
I hope that this addresses your concern.
